static/js/drafts.js

```diff
@@ -454,15 +509,100 @@ export function launch() {
+            return {
+                stream_name,
+                topic: compose_state.topic(),
+                private_recipients: undefined,
+            };
+        } else if (compose_state.get_message_type() === "private") {
+            return {
+                stream_name: undefined,
+                topic: undefined,
+                private_recipients: compose_state.private_message_recipient(),
+            };
+        }
+        return {
+            stream_name: undefined,
+            topic: undefined,
+            private_recipients: undefined,
+        };
+    }
+
+    function filter_drafts_by_compose_box_and_recipient(drafts) {
+        const {stream_name, topic, private_recipients} = current_recipient_data();
+        const stream_id = stream_name ? stream_data.get_stream_id(stream_name) : undefined;
+        const narrow_drafts_ids = [];
+        for (const [id, draft] of Object.entries(drafts)) {
+            // Match by stream and topic.
+            if (
+                stream_id &&
+                topic &&
+                draft.topic &&
+                util.same_recipient(draft, {type: "stream", stream_id, topic})
+            ) {
+                narrow_drafts_ids.push(id);
+            }
+            // Match by only stream.
+            else if (
+                draft.type === "stream" &&
+                stream_id &&
+                !topic &&
+                draft.stream_id === stream_id
+            ) {
+                narrow_drafts_ids.push(id);
+            }
+            // Match by private message recipient.
+            else if (
+                draft.type === "private" &&
+                private_recipients &&
+                _.isEqual(
+                    draft.private_message_recipient
+                        .split(",")
+                        .map((s) => s.trim())
+                        .sort(),
+                    private_recipients
+                        .split(",")
+                        .map((s) => s.trim())
+                        .sort(),
+                )
+            ) {
+                narrow_drafts_ids.push(id);
+            }
+        }
+        return _.pick(drafts, narrow_drafts_ids);
+    }
+
+    const drafts = draft_model.get();
+    const narrow_drafts = filter_drafts_by_compose_box_and_recipient(drafts);
+    const other_drafts = _.pick(
+        drafts,
+        _.difference(Object.keys(drafts), Object.keys(narrow_drafts)),
+    );
+    const formatted_narrow_drafts = format_drafts(narrow_drafts);
+    const formatted_other_drafts = format_drafts(other_drafts);
+
+    render_widgets(formatted_narrow_drafts, formatted_other_drafts);
 
     // We need to force a style calculation on the newly created
     // element in order for the CSS transition to take effect.
     $("#draft_overlay").css("opacity");
 
     open_overlay();
-    set_initial_element(drafts);
+    set_initial_element(formatted_narrow_drafts.concat(formatted_other_drafts));
```

Comment:
I previously was passing the wrong thing here, so it wasn't focusing on anything. I think specifically I was passing the drafts from before they'd gone through `format_drafts`.